File: docker/execute.js
const vm = require('vm');
const util = require('util');

function executeCode(code) {
  const output = [];
  let error = null;

  try {
    // Create a custom console that captures output
    const customConsole = {
      log: (...args) => output.push(args.map(arg => util.inspect(arg)).join(' ')),
      error: (...args) => output.push('ERROR: ' + args.map(arg => util.inspect(arg)).join(' ')),
      warn: (...args) => output.push('WARN: ' + args.map(arg => util.inspect(arg)).join(' ')),
      info: (...args) => output.push('INFO: ' + args.map(arg => util.inspect(arg)).join(' '))
    };

    // Create a sandbox with limited access
    const sandbox = {
      console: customConsole,
      setTimeout: setTimeout,
      setInterval: setInterval,
      clearTimeout: clearTimeout,
      clearInterval: clearInterval,
      require: require, // Limited require for safe packages
      Buffer: Buffer,
      process: { env: {} } // Empty env for security
    };

    // Execute code with timeout
    const script = new vm.Script(code);
    const context = vm.createContext(sandbox);
    
    script.runInContext(context, {
      timeout: 5000, // 5 second timeout
      breakOnSigint: true
    });

  } catch (e) {
    error = {
      type: e.name,
      message: e.message,
      stack: e.stack
    };
  }

  return {
    output: output.join('\n'),
    error: error,
    timestamp: new Date().toISOString()
  };
}

// Read code from stdin
let code = '';
process.stdin.on('data', chunk => {
  code += chunk;
});

process.stdin.on('end', () => {
  const result = executeCode(code);
  console.log(JSON.stringify(result));
});

File: docker/execute.py
#!/usr/bin/env python3
import sys
import io
import contextlib
import json
import traceback
from datetime import datetime

def execute_code(code):
    """Execute Python code in a controlled environment."""
    output = io.StringIO()
    error = None
    result = None
    
    try:
        # Capture stdout
        with contextlib.redirect_stdout(output):
            # Create restricted globals
            restricted_globals = {
                '__builtins__': __builtins__,
                'print': print,
            }
            
            # Execute the code
            exec(code, restricted_globals)
            result = output.getvalue()
            
    except Exception as e:
        error = {
            'type': type(e).__name__,
            'message': str(e),
            'traceback': traceback.format_exc()
        }
    
    return {
        'output': output.getvalue(),
        'error': error,
        'timestamp': datetime.utcnow().isoformat()
    }

if __name__ == '__main__':
    # Read code from stdin
    code = sys.stdin.read()
    
    # Execute and return results
    result = execute_code(code)
    print(json.dumps(result))